### ISSUE TITLE
Question about the Elmish loop

### ISSUE THREAD

**Issue by sheganinans** [fable-compiler/fable-elmish#285]
### Description

I am working on an implementation of an [Elmish loop for Raylib](https://github.com/sheganinans/Elmish.Ray/blob/master/Elmish.Ray/Elmish.Ray.fs#L40), I've got a very basic implementation working however there is one issue. 

The way the [Elmish loop](https://github.com/elmish/elmish/blob/v4.x/src/program.fs#L179) is written, it expects there to be a message in the ring buffer, and if it does not exist, then `processMsg` finishes and exits out of `runWithDispatch`.

The issue currently is that when `runWithDispatch` exits, then the Ray window will close and the program will exit.

### Repro code

[Code Example Here](https://github.com/sheganinans/Elmish.Ray/blob/master/Example/Example.fs#L74).

If you remove line 74, then what will happen is the window will display for a single frame and then will exit.

However if you are continuously finishing the `update` function with some arbitrary message such as `dispatch Continue`, then the loop will continue and the program will continue to run. The window updates and everything works as expected.

### Desired Resolution

So currently I am able to work around the issue by just ensuring that there is always a message to be processed by the Elmish loop with the `dispatch` hack, however I'd like to avoid having to use this `dispatch` hack in `update`.

I'm guessing that I will have to be a bit more clever with how I have implemented [runWithWindow](https://github.com/sheganinans/Elmish.Ray/blob/master/Elmish.Ray/Elmish.Ray.fs#L40). But I'm not exactly sure what I need to do to keep the window opened without the `dispatch` hack.

I'm looking for some guidance on how to properly implement this loop so that I don't have to resort to this `dispatch` hack. 

Thanks!

### Related information 

* elmish version: 4.1.0
* Operating system: Linux

**Comment by et1975** [comment]
As you probably realized "loop" is a bit of a misnomer, since there's no
loop, only event handlers.

You could check other elmish adaptations for inspiration, like Bolero,
Terminal.Gui or Avalonia, but it really depends on the UI framework how you
keep it alive. Or how you correctly implement dispatch and termination, for
that matter.

WPF, for example, like all Windows implementations have their own dispatch
loop for each window, so all we had to do was
https://github.com/elmish/Elmish.WPF/blob/v3/src/Elmish.WPF/Program.fs#L52.

You could also write your own loop, if that makes sense.

**Comment by sheganinans** [comment]
Yes definitely, loop is a bit of a misnomer. So before writing this issue I did take a look at the Avalonia implementation, and now after looking at the Elmish.WPF, Terminal.Gui, and Bolero implementations, I've come to realize that many of us share the same issue. As mentioned in issue #210 I think this issue comes down to having the `program` struct being `private` and needing to be able to split out initialization and the application loop.

I did initially think about implementing my own `runProgram` implementation, and after looking at how the others have done it, I think that is likely the right approach given the current situation. I'll do some experiments and report back. The nice thing is that currently the workaround is very simple!

**Comment by et1975** [comment]
That issue has several unanswered questions, so I can't tell if it really does have overlap with your question. It also predates the subscriptions rework that I believe offers an ideal way for a view implementor to delay/control subscriptions past `init` - ie. the derived program has an option to keep the private model state that would be passed into each "loop" iteration and let the sub to re-eval conditions.

**Comment by sheganinans** [comment]
Definitely! There are still open questions on that route for sure. So I did get working a hacked loop, similar to Bolero's implementation, where I just copied the code for `RingBuffer` and implemented my own `runWithDispatch` which (ab)uses reflection and returns an anonymous function which is called on each frame. I used the code in the 4.x branch and modified it to suit the needs for Ray. 

It's working and I'm pretty happy with it! Even though it's an ugly hack.


There are a few things to note at this current stage: 
* I have the need to be able to update the state on every frame, I'm assuming that I'll be able to easily do that with subscriptions.
* The way that Raylib works, there are two functions `BeginDrawing` and `EndDrawing` ideally I would like to not have to call these in every view, however this interacts poorly with dispatching. If there is a `dispatch` called in between these functions, then the program goes into an infinite loop. I have two solutions:
  * Just call `BeginDrawing` and `EndDrawing` in every view, and ensure all `dispatch`es are done outside of their scope.
  * Introduce a new function to the Elmish "loop" called a "`keyHandler`", which is called outside this scope. Meaning that `view`s never dispatch.

Neither seems exactly ideal, since one demands the library user to call these functions redundantly in every loop *and* to follow an implicit contract about their usage, and the other introduces a new bespoke concept for my implementation.

It seems like `view`s will always be a function of just the model, and will never dispatch, but that is simple enough to enforce by just calling the inner `mkProgram` as:

 `Program.mkProgram init update (fun model _ -> view model)`

However I think there is plenty of room in the design space to allow for staying withing Elmish expectations, we'll see.


Next steps is to rewrite the code to see if I can use subscriptions, that way it follows the intended path of implementation.

More experimentation and more updates sometime later this week! Hopefully I'll be able to get it all working with subscriptions and it should all work seamlessly with the 4.x branch!
